Backend: implement LoadContext

Beanow opened the pull request:
Part of the load refactor #1586
Depends on #1619 and #1620

It uses a new trick of aliasing external module functions as private properties. This makes the spyOn / mock tests more robust, while fitting in the composition responsibility.

Test plan: `yarn test`

Edit: **Conflicts with #1624**
`LoadResult` needs to include the new PluginDeclarations type.

decentralion commented on src/backend/loadContext.js:
```diff
@@ -0,0 +1,115 @@
+    );
+    const plugins = this._declarations(this._pluginLoaders, project);
+    const cred = await this._computeTask(this._compute, this._options, {
+      params: params || {},
```
we shouldn't need `|| {}` -- the type signature for computTask already allows null-ish params

Beanow replied:
I'm afraid not. This is another case where we're mismatching:
`key?: $Shape<T>` (what `.compute()` uses)
`key: ?$Shape<T>` (what `load` uses)

However. It's a good point, I want to fix that at this invocation and move this problem to the `api/load.js` file.

Beanow replied:
Addressed by updating signature here. See https://github.com/sourcecred/sourcecred/pull/1621#issuecomment-582116351

Beanow commented:
Rebased as 1 of 2 dependencies got merged.

Beanow commented:
1) I've updated the LoadResult to include `PluginDeclarations` which addresses the conflict with #1624.
https://github.com/sourcecred/sourcecred/blob/f7c92c9ada14dbc1c0321c9ff6b2e4cf879e58ec/src/backend/loadContext.js#L21-L25

2) To address a nit about optional arguments. I've updated the `LoadContext.load` signature to use named optional parameters. Not only making `TimelineCredParameters` optional, but also the override weights.

https://github.com/sourcecred/sourcecred/blob/f7c92c9ada14dbc1c0321c9ff6b2e4cf879e58ec/src/backend/loadContext.js#L33-L36
https://github.com/sourcecred/sourcecred/blob/f7c92c9ada14dbc1c0321c9ff6b2e4cf879e58ec/src/backend/loadContext.js#L91-L94

3) Finally I've updated tests to cover both changes.
